packages/core/tests/index.test.ts

```diff
@@ -685,4 +685,49 @@ describe('@reatom/core', () => {
       ).toThrowError('[reatom] One of dependencies has the equal id')
     })
   })
+
+  test('should only call the reducer if the dependency state has changed', () => {
+    const increment = declareAction()
+    const initialState = {
+      counter: 0,
+      data: {
+        counter: 1,
+      },
+    }
+    const dataReducerMock = jest.fn(state => state.data)
+    const counterReducerMock = jest.fn(state => state.counter)
+
+    const rootAtom = declareAtom(initialState, on => [
+      on(increment, state => ({ ...state, counter: state.counter + 1 })),
+    ])
+    const dataAtom = map(rootAtom, dataReducerMock)
+    const counterAtom = map(dataAtom, counterReducerMock)
+
+    const store = createStore(counterAtom)
+
+    expect(dataReducerMock).toBeCalledTimes(1)
+    expect(counterReducerMock).toBeCalledTimes(1)
+
+    store.dispatch(increment())
+
+    expect(dataReducerMock).toBeCalledTimes(2)
+    expect(counterReducerMock).toBeCalledTimes(1)
+  })
+
+  test('check equality if returns NaN from atom reducer', () => {
+    const action = declareAction()
+    const rootAtom = declareAtom(0, on => [on(action, () => NaN)])
+
+    const counterReducerMock = jest.fn(state => state)
+    const counterAtom = map(rootAtom, counterReducerMock)
+
+    const store = createStore(counterAtom)
+    expect(counterReducerMock).toBeCalledTimes(1)
+
+    store.dispatch(action())
+    expect(counterReducerMock).toBeCalledTimes(2)
+
+    store.dispatch(action())
+    expect(counterReducerMock).toBeCalledTimes(2)
+  })
```

Comment:
If `hasAtomNewState` and `atomStateNew === atomStateSnapshot` `changedIds` must not contains atom id.
Try this case:
```js
declareAtom(on => [
  on(action, s => s + 1),
  on(action, s => s - 1)
])
```